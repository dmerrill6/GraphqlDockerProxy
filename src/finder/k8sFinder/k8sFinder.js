
//@flow
const Client = require('kubernetes-client').Client
const config = require('kubernetes-client').config
import { FindEndpoints } from '../findEndpointsInterface'
export class K8sFinder extends FindEndpoints{
  constructor(){
    super()
  }


  getEndpoints = async() => {
    const client = new Client({ config: config.fromKubeconfig() })
    await client.loadSpec()
    // const namespaces = await client.api.v1.namespaces.get()
    // console.log(Object.keys(client.api.extension.v1beta1.deployments.get()))
    const services = await client.api.v1.services.get()//client.api.v1.services.get()
    console.log(JSON.stringify(services))
    // namespaces.body.items.forEach((one) => {
    //   console.log('hier bin ich ', JSON.stringify(one))

    // })
    return {}
  }
}

