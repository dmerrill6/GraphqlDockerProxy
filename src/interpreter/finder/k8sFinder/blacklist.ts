let data = []

export const addNamespaceToBlacklist = (name) => {
  data.push(name)
}

export const isNamespaceAtBlacklist = (name) => {
  for (const one in data) {
    if (data[one] === name) {
      return true
    }
  }
  return false
}

export const clearAll = () => {
  data = []
}

export const getBlacklist = () => {
  return data
}
