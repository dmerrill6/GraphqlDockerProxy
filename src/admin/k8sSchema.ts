import { getBlacklist, clearAll } from '../interpreter/finder/k8sFinder/blacklist'

export const typeDefs = `
type Kubernetes{
    blacklist: [String]
    clearBlackList: Boolean
  }
type Query{
    kubernetes: Kubernetes

  }
`

export const resolvers = {
  Query: {
    kubernetes: () => {
      return {
        blacklist: () => {
          return getBlacklist()
        },
        clearBlackList: () => {
          clearAll()
          return true
        },
      }
    },
  },

}
