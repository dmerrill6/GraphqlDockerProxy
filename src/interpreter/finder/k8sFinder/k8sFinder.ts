
const Client = require('kubernetes-client').Client
const config = require('kubernetes-client').config
import * as clientLabels from '../../clientLabels'
import { FindEndpoints } from '../findEndpointsInterface'
import { Endpoints } from '../../endpoints'
import { token, kubernetesConfigurationKind } from '../../../properties'
import { getInClusterByUser } from './getInClusterByUser'
import { addNamespaceToBlacklist, isNamespaceAtBlacklist } from './blacklist'
declare function idx(obj: any, callBack: any):any

export class K8sFinder extends FindEndpoints{
  constructor() {
    super()
  }

  updateUrl = (url:string, sockData:any) :string => {
    if (url.startsWith('http')) {
      return url
    }
    return 'http://' + sockData.metadata.name + '.' + sockData.metadata.namespace + url

  }

  getEndpoints = async(): Promise<Endpoints> => {
    const result = {}
    let client : any = {}
    winston.info('Load K8s')
    switch (kubernetesConfigurationKind()){
      case 'fromKubeconfig': {
        winston.info('Load fromKubeconfig')
        client = new Client({ config: config.fromKubeconfig() })
        break
      }
      case 'getInCluster': {
        winston.info('Load getInCluster')
        client = new Client({ config: config.getInCluster() })
        break
      }
      case 'getInClusterByUser': {
        winston.info('Load getIntClusterByUser')
        client = new Client({ config: getInClusterByUser() })
      }
    }
    await client.loadSpec()
    // const namespaces = await client.api.v1.namespaces.get()
    // console.log())
    const allNamespaces = await client.api.v1.namespaces.get()
    const items = allNamespaces.body.items
    for (const one in items) {
      const namespaceObj = items[one]
      const k8sNamespace = namespaceObj.metadata.name
      if (isNamespaceAtBlacklist(k8sNamespace)) {
        continue
      }
      // const services = client.api.v1.namespaces(namespace).services.get()
      try {
        const services = await client.api.v1.namespaces(k8sNamespace).services.get()
        const servicesItems = services.body.items
        for (const oneService in servicesItems) {
          const oneServiceItem = servicesItems[oneService]
          if (idx(oneServiceItem, _ => _.metadata.annotations[clientLabels.TOKEN]) === token()) {
            const url = this.updateUrl(oneServiceItem.metadata.annotations[clientLabels.URL],
                                       oneServiceItem)
            const namespace = oneServiceItem.metadata.annotations[clientLabels.NAMESPACE]
            if (result[namespace] === undefined) {
              result[namespace] = []
            }
            const deploymentName = oneServiceItem.spec.selector.app
            const deployments = await client.apis.apps.v1beta2.namespaces(k8sNamespace)
                                .deployments.get()
            const compareDeployments = deployments.body.items.filter((one) => {
              return one.spec.template.metadata.labels.app === deploymentName
            })
            let created = ''
            compareDeployments.forEach((one) => {
              created += one.metadata.creationTimestamp
            })
            result[namespace].push({
              url,
              namespace,
              typePrefix: namespace + '_',
              __created: created,
              __imageID: '',
            })
          }
        }
      } catch (e) {
        addNamespaceToBlacklist(k8sNamespace)
        // no loging because user have no permission
        // console.log('error by reading namespace:' + k8sNamespace + ' ', e)
      }

    }
    return result
  }
}
