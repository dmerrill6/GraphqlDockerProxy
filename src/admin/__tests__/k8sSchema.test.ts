import { resolvers, typeDefs } from '../k8sSchema'
import { clearAll } from '../../interpreter/finder/k8sFinder/blacklist'
jest.mock('../../interpreter/finder/k8sFinder/blacklist', () => {
  return {
    getBlacklist: () => {
      return [
        '1',
        '2',
        '3',
      ]
    },
    clearAll: jest.fn(),
  }
})
describe('test k8sSchema', () => {
  it('snapshot the typeDefs', () => {

    expect(typeDefs).toMatchSnapshot()// tslint:disable-line
  })

  it('snapshot the resolver function', () => {
    expect(resolvers).toMatchSnapshot()// tslint:disable-line

  })

  it('tests resolver kubernetes=>blacklist ', () => {

    expect(resolvers.Query.kubernetes().blacklist()).toMatchSnapshot()// tslint:disable-line
  })
  it('tests resolver kubernetes=>clearBlackList ', () => {
    expect(resolvers.Query.kubernetes().clearBlackList()).toBeTruthy()// tslint:disable-line
    expect(clearAll).toBeCalled()

  })
})
