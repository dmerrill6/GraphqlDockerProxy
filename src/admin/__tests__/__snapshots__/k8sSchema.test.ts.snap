// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test k8sSchema snapshot the resolver function 1`] = `
Object {
  "Query": Object {
    "kubernetes": [Function],
  },
}
`;

exports[`test k8sSchema snapshot the typeDefs 1`] = `
"
type Kubernetes{
    blacklist: [String]
    clearBlackList: Boolean
  }
type Query{
    kubernetes: Kubernetes

  }
"
`;

exports[`test k8sSchema tests resolver kubernetes=>blacklist  1`] = `
Array [
  "1",
  "2",
  "3",
]
`;
